### PR TITLE
avro: initialize buffer to avoid potential UB

### LINKED ISSUE #8669: Fix published mz-avro crate on crates.io
Hi,

I think I have found a memory-safety/soundness issue on this crate while reviewing it.

`read_exact` method creates an uninitialized buffer and passes it to user-provided Read implementation. This is unsound, because it may allow safe Rust code to exhibit an undefined behavior (read from uninitialized memory).

[This part](https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read) from the `Read` trait documentation explains the issue:

> It is your responsibility to make sure that buf is initialized before calling read. Calling read with an uninitialized buf (of the kind one obtains via MaybeUninit<T>) is not safe, and can lead to undefined behavior.

This idiom has been referenced in rustsec as UB in the past e.g. [here](https://rustsec.org/advisories/RUSTSEC-2020-0154.html) or [here](https://rustsec.org/advisories/RUSTSEC-2021-0014.html).

The offending line is 

https://github.com/MaterializeInc/materialize/blob/386a0715280d57c02305f5578fe2e46c64cc32c3/src/avro/src/reader.rs#L265

Thanks!

### PR DESCRIPTION
**Pull request by petrosagg**
### Motivation

Fixes #8669

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/9983)

### CODE REVIEW

**Review by umanwizard — APPROVED** [review]

**Comment by benesch** [comment]
Is there no perf impact to this? If so, why was the original code written using `unsafe`? Did we just import it from upstream?

**Comment by umanwizard** [comment]
I really doubt the perf impact of an extra memset will be measurable, but we could try running a benchmark if it's concerning to you @benesch 

To answer your question: yes, we imported this from upstream, and indeed our usage of it is sound (it's not possible to trigger UB from Materialize). But if the perf impact is minimal it's probably best to get rid of the soundness hole, for people's peace of mind.

File: src/avro/src/reader.rs [hunk 1]
```diff
@@ -258,13 +258,9 @@ impl<R: AvroRead> Reader<R> {
     fn fill_buf(&mut self, n: usize) -> Result<(), AvroError> {
         // We don't have enough space in the buffer, need to grow it.
         if n >= self.buf.capacity() {
```
Comment (umanwizard) [inline src/avro/src/reader.rs:260]: I think this needs to be `len` now, not `capacity`.

Comment (petrosagg) [reply src/avro/src/reader.rs]: you're right, fixed

**Pull request merged**